core/driver/sources/src/main/scala/org/virtuslab/ideprobe/ide/intellij/IntelliJProvider.scala

```diff
@@ -0,0 +1,183 @@
+  ): Unit = {
+    val allPlugins = InternalPlugins.probePluginForIntelliJ(version) +: plugins
+
+    case class PluginArchive(plugin: Plugin, archive: Resource.Archive) {
+      val rootEntries: Set[String] = archive.rootEntries.toSet
+    }
+    val targetDir = intelliJ.paths.plugins
+    val archives = withParallel[Plugin, PluginArchive](allPlugins)(_.map { plugin =>
+      val file = dependencies.fetch(plugin)
+      PluginArchive(plugin, file.toArchive)
+    })
+
+    val distinctPlugins = archives.reverse.distinctBy(_.rootEntries).reverse
+
+    parallel(distinctPlugins).forEach { pluginArchive =>
+      pluginArchive.archive.extractTo(targetDir)
+      println(s"Installed ${pluginArchive.plugin}")
+    }
+  }
+
+  private def withParallel[A, B](s: Seq[A])(f: JStream[A] => JStream[B]): Seq[B] = {
+    f(parallel(s)).collect(Collectors.toList[B]).asScala.toList
+  }
+
+  private def parallel[A](s: Seq[A]) = {
+    s.asJava.parallelStream
+  }
+}
+
+final class ExistingIntelliJ(
+    dependencies: DependencyProvider,
+    path: Path,
+    override val plugins: Seq[Plugin],
+    override val paths: IdeProbePaths,
+    override val config: DriverConfig
+) extends IntelliJProvider {
+  override val version = IntelliJVersionResolver.version(path)
+  override def withConfig(config: DriverConfig): ExistingIntelliJ =
+    new ExistingIntelliJ(dependencies, path, plugins, paths, config)
+
+  override def withPaths(paths: IdeProbePaths): ExistingIntelliJ =
+    new ExistingIntelliJ(dependencies, path, plugins, paths, config)
+
+  override def withPlugins(plugins: Plugin*): ExistingIntelliJ =
+    new ExistingIntelliJ(dependencies, path, this.plugins ++ plugins, paths, config)
+
+  override def setup(): InstalledIntelliJ = {
+    val intelliJPaths = IntelliJPaths.fromFile(path)
+    val pluginsDir = intelliJPaths.plugins
+    val backupDir = Files.createTempDirectory(path, "plugins")
+    val intelliJ = new LocalIntelliJ(path, paths, config, intelliJPaths, backupDir)
+    pluginsDir.copyDir(backupDir)
+    installPlugins(dependencies, plugins, intelliJ)
+
+    intelliJ
+  }
+}
+
+final class IntelliJFactory(
+    dependencies: DependencyProvider,
+    override val plugins: Seq[Plugin],
+    override val version: IntelliJVersion,
+    override val paths: IdeProbePaths,
+    override val config: DriverConfig
+) extends IntelliJProvider {
+  override def withConfig(config: DriverConfig): IntelliJFactory =
+    new IntelliJFactory(dependencies, plugins, version, paths, config)
+
+  override def withPaths(paths: IdeProbePaths): IntelliJFactory =
+    new IntelliJFactory(dependencies, plugins, version, paths, config)
+
+  override def withPlugins(plugins: Plugin*): IntelliJProvider =
+    new IntelliJFactory(dependencies, plugins = this.plugins ++ plugins, version, paths, config)
+
+  override def setup(): InstalledIntelliJ = {
+    val root = createInstanceDirectory(version)
+
+    val intelliJPaths: IntelliJPaths =
+      IntelliJPaths(
```

Comment:
oh, like this...
I thought it would be another static method for `IntelliJPaths` that would hide this details, and possibly share some code with the other method. For example 
`IntelliJPaths.fromExistingInstance(root)` or `IntelliJPaths.fromExistingPreferences(root.resolve("bin/idea.prefs"))`
and
`IntelliJPaths.default(root)`
or something like this

Reply:
Done.